aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1719 @@
+    /// Therefore, should charge delegator the maximum amount it would unfairly earn only this epoch.
+    public fun get_add_stake_fee(pool_address: address, amount: u64): u64 {
+        if (stake::is_current_epoch_validator(pool_address)) {
+            let (rewards_rate, rewards_rate_denominator) = staking_config::get_reward_rate(&staking_config::get());
+            if (rewards_rate_denominator > 0) {
+                ((((amount as u128) * (rewards_rate as u128)) / (rewards_rate_denominator as u128)) as u64)
+            } else { 0 }
+        } else { 0 }
+    }
+
+    /// Add `amount` of coins to the delegation pool `pool_address`.
+    public entry fun add_stake(delegator: &signer, pool_address: address, amount: u64) acquires DelegationPool {
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+        // short-circuit if amount to add is 0 so no event is emitted
+        if (amount == 0) { return };
+
+        let pool = borrow_global_mut<DelegationPool>(pool_address);
+        let delegator_address = signer::address_of(delegator);
+
+        // fee to be charged for adding `amount` stake on this delegation pool at this epoch
+        let add_stake_fee = get_add_stake_fee(pool_address, amount);
+
+        // stake the entire amount to the stake pool
+        coin::transfer<AptosCoin>(delegator, pool_address, amount);
+        stake::add_stake(&retrieve_stake_pool_signer(pool), amount);
+        // but buy shares for delegator just for the remaining amount after fee
+        buy_in_active_shares(pool, delegator_address, amount - add_stake_fee);
+
+        // commit coins from `add_stake` fee to the active shares pool
+        // in order to not mistake them for epoch active rewards
+        let (active, _, pending_active, _) = stake::get_stake(pool_address);
+        pool_u64::update_total_coins(&mut pool.active_shares, active + pending_active);
+
+        event::emit_event(
+            &mut pool.add_stake_events,
+            AddStakeEvent {
+                pool_address,
+                delegator_address,
+                amount_added: amount,
+                add_stake_fee,
+            },
+        );
+    }
+
+    /// Unlock `amount` from the active + pending_active stake of `delegator` or
+    /// at most how much active stake there is on the stake pool.
+    public entry fun unlock(delegator: &signer, pool_address: address, amount: u64) acquires DelegationPool {
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+        // short-circuit if amount to unlock is 0 so no event is emitted
+        if (amount == 0) { return };
+
+        let pool = borrow_global_mut<DelegationPool>(pool_address);
+        let delegator_address = signer::address_of(delegator);
+
+        // capture how much stake would be unlocked on the stake pool
```

Comment:
I think we should fail in the case someone tries to unlock more than what's remaining active in the stake pool. Silently lowering the amount can be confusing at unlock time. We've seen similar confusion due to a similar logic in staking_contract::withdraw and I think it's even more confusing if we do this here in unlock.